### ISSUE TITLE
[Bug] no-direct-mutation-state docs example shows this.state.foo = X but rule doesn't detect it

### ISSUE THREAD

**Issue by xcwdss** [Rel1cx/eslint-react#1817]
## Summary                                                                                                                                                                                                        
                                                                                                                                                                                                                    
  The documentation for [`no-direct-mutation-state`](https://eslint-react.xyz/docs/rules/no-direct-mutation-state) shows an example asserting that `this.state.foo = "baz"` will be reported, but the rule          
  implementation only detects whole assignment `this.state = X`. Docs and source have drifted.                                                                                                                      
                                                                                                                                                                                                                    
  ## Reproduction                                                                                                                                                                                                   
                                                                                                                                                                                                                    
  `reproduce.tsx`:

  ```tsx                                                                                                                                                                                                            
  import React from 'react'
                                                                                                                                                                                                                    
  class MyComponent extends React.Component<{}, { foo: string }> {                                                                                                                                                
    state = { foo: 'bar' }                                                                                                                                                                                          
    render() {                                                                                                                                                                                                      
      return (                                                                                                                                                                                                      
        <div                                                                                                                                                                                                        
          onClick={() => {                                                                                                                                                                                          
            this.state.foo = 'baz'  // ← docs say this should be reported                                                                                                                                           
          }}                                                                                                                                                                                                      
        >                                                                                                                                                                                                           
          {this.state.foo}                                                                                                                                                                                        
        </div>                                                                                                                                                                                                      
      )
    }                                                                                                                                                                                                               
  }                                                                                                                                                                                                               
  ```                                                                                                                                                                                                             

  `eslint.config.mjs`:                                                                                                                                                                                              
  
  ```js                                                                                                                                                                                                             
  import reactX from '@eslint-react/eslint-plugin'                                                                                                                                                                
  export default [                                                                                                                                                                                                  
    reactX.configs['recommended-typescript'],
    { rules: { '@eslint-react/no-direct-mutation-state': 'error' } },                                                                                                                                               
  ]                                                                                                                                                                                                                 
  ```                                                                                                                                                                                                               
                                                                                                                                                                                                                    
  ```bash                                                                                                                                                                                                         
  $ npx eslint reproduce.tsx
  # (no output — rule didn't report)                                                                                                                                                                                
  ```                                                                                                                                                                                                               
                                                                                                                                                                                                                    
  ## Expected (per docs)                                                                                                                                                                                            
                                                                                                                                                                                                                  
  The official docs example explicitly comments:                                                                                                                                                                    
  
  ```tsx                                                                                                                                                                                                            
  this.state.foo = "baz";                                                                                                                                                                                         
  //   ^^^ Do not mutate state directly. Use 'setState()' instead.                                                                                                                                                  
  ```                                                                                                                                                                                                               
                                                                                                                                                                                                                    
  So the rule **should** report this line.                                                                                                                                                                          
                                                                                                                                                                                                                  
  ## Actual

  The rule does **not** report `this.state.foo = "baz"`. It only reports whole-state replacement like `this.state = { foo: 'baz' }`.                                                                                
  
  ## Root cause (source analysis)                                                                                                                                                                                   
                                                                                                                                                                                                                  
  In `plugins/eslint-plugin-react-x/src/rules/no-direct-mutation-state/no-direct-mutation-state.ts`, the rule calls `core.isAssignmentToThisState(node)`. That helper, defined in `@eslint-react/core`:             
                                                                                                                                                                                                                  
  ```ts                                                                                                                                                                                                             
  function isAssignmentToThisState(node) {                                                                                                                                                                        
    const { left } = node;                                                                                                                                                                                          
    return left.type === 'MemberExpression'                                                                                                                                                                         
      && left.object.type === 'ThisExpression'   // ← strict: left.object must be `this` itself                                                                                                                     
      && getPropertyName(left.property) === 'state';                                                                                                                                                                
  }                                                                                                                                                                                                                 
  ```                                                                                                                                                                                                               
                                                                                                                                                                                                                    
  For `this.state.foo = "baz"`:                                                                                                                                                                                     
  - `node.left` = `MemberExpression(this.state, .foo)`                                                                                                                                                            
  - `node.left.object` = `MemberExpression(this, .state)` ← this is a `MemberExpression`, **not** a `ThisExpression`                                                                                                
  - Helper returns `false` → rule skips → not reported.                                                                                                                                                             
                                                                                                                                                                                                                    
  The helper only matches `this.state = X` where `left.object === ThisExpression`.                                                                                                                                  
                                                                                                                                                                                                                    
  ## Verified on multiple versions                                                                                                                                                                                  
                                                                                                                                                                                                                  
  | Version | helper behavior | rule behavior |                                                                                                                                                                     
  |---|---|---|
  | 1.53.1 | matches only `this.state = X` | reports only `this.state = X` |                                                                                                                                        
  | 5.8.8 (latest) | identical, with `@deprecated` JSDoc added | identical |                                                                                                                                        
  | main branch | identical | identical |                                                                                                                                                                           
                                                                                                                                                                                                                    
  ## Comparison: classic `eslint-plugin-react`                                                                                                                                                                      
                                                                                                                                                                                                                  
  [`eslint-plugin-react/lib/rules/no-direct-mutation-state.js`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/no-direct-mutation-state.js) uses `getOuterMemberExpression(node.left)`     
  which descends `.object.object...` to find the innermost `this.state`, correctly reporting any depth of mutation.
                                                                                                                                                                                                                    
  ## Suggested fix                                                                                                                                                                                                  
  
  Two options:                                                                                                                                                                                                      
                                                                                                                                                                                                                  
  1. **Update the helper** to descend the MemberExpression chain, matching what the docs promise (closer to classic behavior):                                                                                      
  
     ```ts                                                                                                                                                                                                          
     function isAssignmentToThisState(node) {                                                                                                                                                                     
       let cur = node.left                                                                                                                                                                                          
       while (cur && cur.type === 'MemberExpression') {                                                                                                                                                           
         const obj = cur.object                                                                                                                                                                                     
         if (                                                                                                                                                                                                       
           obj?.type === 'MemberExpression' &&
           obj.object?.type === 'ThisExpression' &&                                                                                                                                                                 
           getPropertyName(obj.property) === 'state'                                                                                                                                                              
         ) return true                                                                                                                                                                                              
         cur = obj
       }                                                                                                                                                                                                            
       // Also catch `this.state = X`                                                                                                                                                                             
       return node.left.type === 'MemberExpression'
         && node.left.object.type === 'ThisExpression'
         && getPropertyName(node.left.property) === 'state'                                                                                                                                                         
     }
     ```                                                                                                                                                                                                            
                                                                                                                                                                                                                  
  2. **Or update the docs** to show only the actually-detected case (`this.state = X`) and remove the misleading shallow-mutation example.                                                                          
  
  Option 1 matches the documented intent and historical classic plugin behavior. Option 2 documents the truth but reduces the rule's value.                                                                         
                                                                                                                                                                                                                  
  ## Environment                                                                                                                                                                                                    
                                                                                                                                                                                                                  
  - `@eslint-react/eslint-plugin`: 1.53.1 (also tested 5.8.8 and main)                                                                                                                                              
  - ESLint: 9.x flat config
  - Node: 20.19.5                                                                                                                                                                                                   
  - OS: macOS Darwin 24.6.0

**Comment by Rel1cx** [comment]
Hi @xcwdss, thank you for the detailed report. The issue will be fixed in the next release. `isAssignmentToThisState` now descends the `MemberExpression` chain and unwraps TypeScript type expressions, so `this.state.foo = "baz"` and `(this.state as any).foo = "baz"` are correctly detected alongside `this.state = X`.
